text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,565 @@
+We could add a sticky flag to `Cargo.lock` though that could also be confusing, see "Configuring the resolver mode on the command-line or `Cargo.toml`".
+
+This would also error or pick lower versions more than it needs to when a workspace contains multiple MSRVs.
+We'd want to extend the resolver to treat Rust as yet another dependency and turn `package.rust-version` into dependencies on Rust.
+This could cause a lot more backtracking which could negatively affect resolver performance for people with lower MSRVs.
+
+If no `package.rust-version` is specified,
+we wouldn't want to fallback to the version of rustc being used because that could cause `Cargo.lock` churn if contributors are on different Rust versions.
+
+# Prior art
+[prior-art]: #prior-art
+
+- Python: instead of tying packages to a particular tooling version, the community instead focuses on their equivalent of the [`rustversion` crate](https://crates.io/crates/rustversion) combined with tool-version-conditional dependencies that allow polyfills.
+  - We have [cfg_accessible](https://github.com/rust-lang/rust/issues/64797) as a first step though it has been stalled
+  - These don't have to be mutually exclusive solutions as conditional compilation offers flexibility at the cost of maintenance.  Different maintainers might make different decisions in how much they leverage each
+  - One big difference is Python continues to support previous releases which sets a standard within the community for "MSRV" policies.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+The config field is fairly rought
+  - The location (within `build`) needs more consideration
+  - The name isn't very clear
+  - The values are awkward
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Improve the experience with lack of `rust-version`
+
+The user experience for this is based on the extent and quality of the data.
+Ensuring we have `package.rust-version` populated more often (while maintaining
+quality of that data) is an important problem but does not have to be solved to
+get value out of this RFC and can be handled separately.
+
+~~We could encourage people to set their MSRV by having `cargo new` default `package.rust-version`.~~
+However, if people aren't committed to verifying it,
+it is likely to go stale and will claim an MSRV much older than what is used in practice.
+If we had the hard-error resolver mode and
+[clippy warning people when using API items stabilized after their MSRV](https://github.com/rust-lang/rust-clippy/issues/6324),
+this will at least annoy people into either being somewhat compatible or removing the field.
+
+~~When missing, `cargo publish` could inject `package.rust-version` using the version of rustc used during publish.~~
+However, this will err on the side of a higher MSRV than necessary and the only way to
+workaround it is to set `CARGO_BUILD_RESOLVER_PRECEDENCE=maximum` which will then lose
+all other protections.
```

Comment:
A similar alternative I would like to use is to explicitly specify the rust-version during publish, `cargo publish --rust-version 1.74`. Currently for crates tracking latest stable only there are three options I see, none of which have been acceptable enough for me to start adding `rust-version` to my crates:

 * Add the `rust-version` then `cargo publish --allow-dirty`; breaks the direct link between package and repository
 * Add the `rust-version`, push to a separate branch/tag, then `cargo publish`; leaves a lot of offshoots in the repository, is more annoying to integrate with checking CI on the commit before publish
 * Run a 6-weekly CI job to update the `rust-version` in the repo; results in a lot of PR spam for low-activity projects

Having this instead be applied to the generated `Cargo.toml` (ideally with an additional comment noting how it was injected) while leaving the `Cargo.toml.orig` matching the repository would be preferable.

Basing it off the currently used `rustc` version would be acceptable, but I would like confirmation in that case, it would be too easy to accidentally publish with an older stable or a nightly and correcting this metadata is annoying.

> However, this will err on the side of a higher MSRV than necessary...

For some crates this _is_ the advertised MSRV policy though. If that causes issues for downstreams that didn't realise they were using something with such a policy, that's a good thing.

Reply:
> Run a 6-weekly CI job to update the rust-version in the repo; results in a lot of PR spam for low-activity projects

This is what cargo does
- https://github.com/rust-lang/cargo/pull/12381
- https://github.com/rust-lang/cargo/pull/12775

> A similar alternative I would like to use is to explicitly specify the rust-version during publish, cargo publish --rust-version 1.74. 

Should we expect users to type it in on the command-line correctly or should we add a meta-value like `rust-version = "current"` that gets populated by `rustc --version` on publish?

Granted, this would be relegated to the Future Possibilities section.  Might not need a full RFC on its own but could potentially be an MCP in cargo.

> For some crates this is the advertised MSRV policy though. If that causes issues for downstreams that didn't realise they were using something with such a policy, that's a good thing.

imo the problem is that its implicit, that the author isn't opting into it and acknowledging it as a policy. 

I like the idea of separate fields because it bypasses all conversations on the quality of the data.  The RFC then intentionally leaves room for us to improve our heuristics for the no-rust-version case over time so we can play with and adjust as we learn from experience.  Storing a heuristic in a user-provided field that is immutable makes that a lot harder.

Reply:
> This is what cargo does

Yeah, it works for cargo since it is high-activity. I have crates which I touch about once every 6 months, having 4 PRs changing the `rustc-version` in that time is... actually not as bad as I was thinking it would be, but still not as good as just injecting the correct value during publish since it's not something that is important to have in-repo.

> imo the problem is that its implicit, that the author isn't opting into it and acknowledging it as a policy.

My crates are very explicit about this:

> #### Rust Version Policy
> This crate only supports the current stable version of Rust, patch releases may use new features at any time.

Reply:
Ah, having `rust-version = "current"` in the `Cargo.toml` would be fine with me too. Though, I'd still require that to have an explicit "Publish with `rust-version = "1.74"`?" question during `cargo publish` to avoid accidentally publishing the wrong version.

I was thinking about the ability to include policy metadata for the later "establish policy on MSRV" section. The data that mentions like "A report of `package.rust-version` for the latest versions of packages on crates.io" feels like it's just trying to infer policies from the lossy available data. Being able to explicitly specify `current - 3` or `6 months ago` would make it feasible to actually work out what the situation will be going forward (if how that relates to semver is also worked out).

Reply:
> "Publish with rust-version = "1.74"?" question during cargo publish to avoid accidentally publishing the wrong version (and an extra warning if the current cargo is a nightly build).

For myself, I'm not to keen on the idea of confirmation prompts in publish.

> The data that mentions like "A report of package.rust-version for the latest versions of packages on crates.io" feels like it's just trying to infer policies from the lossy available data. 

To be clear, the goal isn't to set policy but to give people data, with appropriate caveats, that they can set their policies.

> Being able to explicitly specify current - 3 or 6 months ago would make it feasible to actually work out what the situation will be going forward (if how that relates to semver is also worked out).

Unsure what would be explicitly specifying "current -3" in this statement and how it ties into the "establish policy on MSRV" section.

Reply:
4fc3608  added a `rust-version-policy` field as at least one way of doing this.

Reply:
Why not just leave the rust version at the lower value if you're not making changes (and therefore presumably wouldn't actually require the new rust version). Wouldn't have `rust-version = current` option lead to artificially inflated minimum rust versions, that will prevent projects from using older compiler even though the code will actually compile just fine with those versions.

Reply:
I feel like we ought to separate the two concepts of:

- A projects MSRV *policy*
- A release's actual MSRV (that it will successfully and correctly compile with)

And allow for the possibility that a release has a much lower MSRV than the projects MSRV policy guarantees.

Reply:
So first off, this is all in future possibilities, so none of this is blocking for this RFC.

That said, I feel like there are two approaches to MSRV policies
- Report the true minimum MSRV in `package.rust-version`, updating it reactively / on-demand
- Proactively update the MSRV to the version specified in a policy

When reactively updating MSRV, projects have different standards set for when this is justified.  For me as a contributor, this can be discouraging as I feel like if I want to do anything its automatically going to be an argument.    I also found with clap 2->3 that people assume your MSRV policy based on your MSRV and they can get very frustrated when their expectations are not continued to be met.

For these reasons, I personally suggest people follow a proactive model.  This gives implicit permission for contributors to new std features, new dependency features, or even new dependencies.  This also makes the MSRV policy clear so there are no disconnects in expectations  I recognize this won't be for everyone.  As we write more documentation to help people create their own MSRV policies, I think it is worthwhile to discuss the trade offs and leave it to them which they should use.

Reply:
My concern is that this will end up pushing the entire ecosystem onto “latest stable” (or alternatively into using the ignore flag for this feature if they’re not), because any time a crate with a “latest stable” MSRV publishes an update (even it’s a minor bug fix that in no way affects MSRV), the MSRV for that crate will be bumped, and that bump will be enforced by cargo. This will be particularly bad for small patch releases.

It could be possible to resolve an older version in such cases, but IMO that’s *worse*, as it means that anybody not using the latest rust version could be missing out on important bug fixes. And they wouldn’t even know that they weren’t getting them.

In my crates I have a policy of never bumping MSRV in patch releases, so this would be much less of an issue. But I can see that clashing with a package like serde’s versioning system where every release is a patch release.

Reply:
> It could be possible to resolve an older version in such cases,

Doing that is the entire point of this RFC

> And they wouldn’t even know that they weren’t getting them.

The RFC calls out having `cargo update` tell users.  Josh has another thread that might end up exploring alternative forms of reporting

Reply:
It would be nice to have `rust-version = "current"`, but also being able to override it while publishing, so you could bisect rustc versions to find the actual minimum supported version.

Even better would be a tool to quickly determine the MSRV, e.g. by checking `#[stable]` attributes of all the used stdlib items, and looking for newer syntax like GATs.

Reply:
`cargo msrv` can help you bisect your MSRV

As for thorough enough static analysis for declaring an MSRV, that is likely to be too large of an effort to take on.  At best we can likely get static analysis to help developers find common MSRV problems earlier in their development process than CI.

Reply:
If there is a "current" shorthand, I think it would be good to provide slightly smarter semver-aware variants "current-major" and "current-minor" that only bump the version when publishing a major or minor versions. Bumping MSRV in patch versions seem like bad practice to me, as it forces people into upgrading Rust in order to get critical bug fixes.

> I do not test with anything except the latest stable, so I cannot be sure what the required lower value actually is.

To me, this sort of approach largely undermines the value of this feature, which is dependent on the MSRVs actually being accurate. If crates just put "latest stable at time of publishing" then they'll be too many false positives in the incompatibility detection for it to be useable.

It's pretty easy to add a CI task to test that compilation still works with the set MSRV version, and then you can either bump the version or remove the offending feature if it fails. I feel like we should at least be encouraging this sort of workflow, even if we don't enforce it.

As an example: a crate I maintain - `taffy` - has a "latest stable" MSRV *policy*, because we don't want to block use of new features where they are useful. But Taffy actually compiles on Rust versions down to `1.65`, because we don't actually need anything newer than that. It would not be great if users of older versions of Rust were blocked from upgrading to new bugfix releases of Taffy simply because we had our MSRV set to "current" when publishing them. It would mean that all users of Taffy would have to use latest stable Rust (or wade through the changelog to every time any transitive dependency publishes an update - but I feel like that's not really feasible in practice), even though it might be the case that none of their dependencies actually require it. And if that's the case, then there would be no need for this feature. They could just use latest stable rust and be done with it.

> I also found with clap 2->3 that people assume your MSRV policy based on your MSRV and they can get very frustrated when their expectations are not continued to be met.

Well if users reacted like that to an MSRV bump in a major version, imagine how they will react when there are MSRV bumps in patch versions!

Reply:
> it simply requires them to acknowledge that they are doing so by using --precise or --ignore-rust-version when pulling those dependencies in.

To add to that list, if you set your version requirement higher (directly or via `cargo add --ignore-rust-version`) then you can also bypass this though the discussion at https://github.com/rust-lang/rfcs/pull/3537#discussion_r1410737006 could affect how this works.